Enhanced PrettyTabBar

echoz opened the pull request:
This is a huge ass commit. Really sorry for the multiple commits. Should've rebased it first.

That aside, this pull request basically is about an enhanced UITabBar. It has the following features,
1. Shadows that can be configured to appear upwards or downwards
2. Switchable representation of the buttons that appear in the TabBar either between the Apple defaults (UITabBarButton) or the PrettyKit highly configurable version (PrettyTabBarButton)

This new TabBar is should be a drop in replacement while supporting all callbacks that is expected of a UITabBar. Has been tested with UITabBarControllers to work perfectly. No changes needed to make this work with existing projects, just use PrettyTabBar instead of an UITabBar.

The new class PrettyTabBarButton itself provides more configuration than previously possible for normal UITabBars while still support iOS 4.3.

Some of the configurable options include,
- Title font
- Title shadow offset and opacity
- Title color
- Badge border color
- Badge font
- Badge text color
- Badge inner gradient start and end colors
- Image to use for the highlight in the select state or;
- Gradient start and end colors for the highlight in select state
- Badge shadow offset and opacity
- Image to use when the button is selected or;
- Gradient start and end colors for the "tint" on the image to use when button is selected

While these options are configurable on the TabBarButton, they are really only settable via usage of the PrettyTabBar as there is tight coupling between both classes (also how Apple seems to do it).

PrettyTabBarButton draws everything using Core Graphics.

Also added some new methods to the PrettyDrawing class.

echoz committed:
added more generic shadow drawing

echoz committed:
alignment for default shadow opacity

echoz committed:
pretty tab bar now draws an upwards and downwards shadow

echoz committed:
renamed selector to fit camel case

echoz committed:
don't draw if opacity is 0.

echoz committed:
initial skeleton for PrettyTabBarButton

echoz committed:
Set a mode to use PrettyTabBarButtons instead of UITabBarButtons.

This uses an internal representation as well as reflecting it in the view hierarchy to allow more configurable tab bar buttons.

echoz committed:
added PrettyTabBarButton class

echoz committed:
some text drawing options plus drawing of text in the correct spot

echoz committed:
if badge value changes, we should update the view's badge value too.

echoz committed:
observe badgeValue and update accordingly.

also corrected logic flow for selection of tabbarbuttons

echoz committed:
draw selection background. updated highlighted text to one that's more obvious.
setting of badge value does the correct things by copying as well as setting that a redraw is needed.

echoz committed:
initial drawing of badges

echoz committed:
setup options for badge

echoz committed:
draws badges according to font size.

echoz committed:
adjusted opacity values. added todos

echoz committed:
default values for text colors and highlights

echoz committed:
activate pretty tab bar buttons

echoz committed:
customize badge text color and also working gradients for badge

echoz committed:
CF like method to return a mutable path that is in the shape of a rounded rectangle

corresponding methods to draw rounded rectangles with a solid fill color and with a gradient.

abstracted gradient drawing into a method that will draw to a context you specify. also used a more primitive way of dealing with the colors

UIColor additions to deal with getting of converted RGB components in the absence of colorspace conversion mechanics

echoz committed:
make sure when switching, buttons that are selected will be in the correct state

echoz committed:
more options for customization

echoz committed:
tightly coupled interfaces to deal with selection of tab bar
also modified font style to match defat UITabBarButtons'

echoz committed:
internal method to handle resetting of selections of prettytabbarbuttons

echoz committed:
documentation for PrettyTabBarButton

echoz committed:
switched to using a UIControl subclass. Clearer separation between Views and Controllers.

echoz committed:
set custom prettytabbarbutton properties

echoz committed:
set correct alignment for assignment of button properties

removed sending super observevalueforkeypath. technically not needed.

echoz committed:
draw gradient tint using tint gradient colors specified.

echoz committed:
use proper sizes when rendering images

echoz committed:
proper memory management

echoz committed:
added configuration for the highlighted image's gradient start and end colors as well as an array to specify the highlighted images (if any)

echoz commented:
Okay. There is a lot of power inherent in its customizability. I made a small example without screwing around too much with it.

![PrettyTabBar Enhanced Example](http://f.cl.ly/items/1t2W04043I1v1D3s392h/Screen%20Shot%202012-05-06%20at%209.17.51%20PM.png)

There is still much to do, like adding the glossy effect and all that, but this should be a good jumping off point.

vicpenap commented:
Wow, this looks great! I'm really looking forward to trying it. Probably tomorrow. In the meanwhile, you could do a more detailed example of use if you want to :). 

Thanks a lot!

echoz commented:
Sure thing. Will probably work on that examples much later. Its been a long weekend. ;)

vicpenap commented:
Of course! No hurries at all :)

echoz committed:
more logical layout of code. helps with memory management of observation status

code to add buttons to the view are done at the switch. layoutSubViews does it namesake, it lays things out and also sets properties.

echoz committed:
forgot to check for if items count is equal to 0 here.

also checked to make sure the index returned is not NSNotFound

echoz committed:
forgot to check for if items count is equal to 0 here.

also checked to make sure the index returned is not NSNotFound

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

echoz committed:
forgotten to intercept setting of items and rebuilding the representation

also more stringent checks on the count of objects in an array

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

echoz commented:
Terribly sorry for the additional commits. realised the way i was laying out subviews wasn't ideal. so rewrote the implementation. this should be much better :)

These should be the last commits.

echoz committed:
forgotten to free up observer

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

vicpenap commented:
Hi!

I've merged your changes into a new local branch, but it isn't working as expected. I've just launched the example app, with no changes, and it has crashed in line 230 of `PrettyTabBar.m`: `[item removeObserver:self forKeyPath:@"badgeValue"];`

I've commented that line, just to make it run, but there are no tabbar buttons, as seen in the included screenshot. 

I see you're still working on the feature, so I'll wait until you're finished :)

![](http://cl.ly/2w3W0V3C01141p1T1s0l/Screenshot_1.png)

echoz committed:
fixed bug where badge value being removed from initial uitabaritems when switching causes crashes

also fixed issues with pretty buttons not appearing

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

echoz commented:
Hello hello. fixed the bug. sorry about that. :/

vicpenap commented:
Great!

It works really great. But, when setting a predefined item from Interface Builder, it won't show.

But creating a custom one works great, and this work you're doing looks very powerful. It would be great, by the way, to tint images in not selected items with the same gray level of the text.

Another thing: I think it would be a good idea to remove default values from PrettyTabBarButtons because, as far as I know, they're always set from PrettyTabBar. This will remove duplicity, and it will be faster to change those default values in the future :)

Last one: I've changed this default values to better mimic the default UITabBar:
- default_text_color                              [UIColor colorWithWhite:0.5 alpha:1.0]
- default_highlight_gradient_start_color          [UIColor colorWithWhite:0.35 alpha:1.0] 
- default_highlight_gradient_end_color            [UIColor colorWithWhite:0.2 alpha:1.0]

echoz commented:
Yeah. Tinting was a problem to figure out. Will probably commit some other changes at a later date, but this should be ready for use.

Yeah, I was also thinking about all the default values in the PrettyTabBarButton. Just added it to be consistent with the project's settings.

Not too sure what you mean by setting predefined item that won't show in interface builder

vicpenap commented:
A picture is worth a thousand words :)

![](http://f.cl.ly/items/3b2W2n2p3x2H0e3m342o/Captura%20de%20pantalla%202012-05-07%20a%20las%2010.52.36.png)

echoz commented:
Ahh! Hmm. That's quite a challenge to get working. I'd probably work on that next weekend. (PrettyKit has been a fun 2 weekends of hacks. :))

But yeah, I guess custom stuff for now.

vicpenap commented:
Hahaha yeah, I'd love to have more time to spend on it too. 

I've pushed a new branch called tabbar, with your changes included, and some changes I've made on yours :D. Including removing the default values from tabbar button, a little bit of refactoring, and tinting the item's icon to the same text color when not selected.

I won't merge it with the master branch yet, as I think we can improve it even more :D. But I'm really looking forward to releasing a new version of PrettyKit with this nice tabbar :D

Cheers!

vicpenap commented:
Well, I forgot to add those changes :). I'm doing it right now.

echoz commented:
That'll be really exciting to work on. :)

Lets make this really Pretty (pun :P)

vicpenap committed:
Some changes to PrettyTabBar stuff

- Changed some default values
- Removed unused default values from PrettyTabBarButton
- Refactored some code in PrettyTabBarButton
- Added tint effect to non-selected tabbar images

vicpenap commented:
Done!

I can't make badges work, though :/

echoz commented:
Weird. I've gotten them to work by modifying the View Controller's UITabBarItem's badgeValue.

PrettyTabBar listens for changes to those and changes the value accordingly.

vicpenap commented:
I've tried, before my changes, that way and also through the IB's field, and it didn't seem to work :/. I haven't tried hard, though.

echoz commented:
Weird. Works for me.

Insert this at line 34 of the appdelegate:  
`((UIViewController *)[self.tabBarController.viewControllers objectAtIndex:0]).tabBarItem.badgeValue = @"1";`

vicpenap commented:
I did that from the view controller (`self.tabBarItem`), that's why it didn't work for me.

echoz commented:
Ah! Okay. I think I know what to do for this. :)

echoz committed:
set item's badge value when laying out subviews.

optionally aggressively passes on KVO observation to super class through the use of a @try..@catch block to prevent exceptions.

echoz commented:
i just checked apple's implementation, it also doesn't seem to reflect the change in the badge value.

Further investigation says that if you try to set the badgeValue during viewDidLoad, it doesn't seem to work. However anytime else after the UIViewController has loaded, calling self.tabBarItem.badgeValue = something should work fine and update the tabbar.

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

echoz commented:
Regarding [#13](https://github.com/vicpenap/PrettyKit/issues/13):

Translucent tabbars would only probably work if they are not part of the UITabBarController as the content view of the tabbar controller will not resize to encompass the area lost (a la translucent UINavigationBar).

Don't really see a point beyond the fact that this might introduce bad UI design.

Btw, added some more commits previously. Not sure if you've merged them.

vicpenap commented:
I've just merged them.

echoz committed:
fixes issue where sometimes buttons don't have an image or title.

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

vicpenap commented:
Hi @echoz. Happy to see tabbar is still alive :D. Do you have any clue about the issue with predefined IB buttons? I mean this: https://github.com/vicpenap/PrettyKit/pull/12#issuecomment-5546144.

echoz commented:
I can't give a definitive answer (because I haven't really had the time to look into it), but as of this moment I don't think its possible to use predefined buttons for this :/

vicpenap commented:
Well, as a last resort we could add a note in documentation to explain that.

echoz commented:
Yeah. I guess that'll do for now. Will look into it when I have more time on my hands.

echoz committed:
bump pretty tab bar button text up by 2 px

echoz committed:
set corner radius for highlights

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

echoz committed:
work around to stop instruments complaining about a memory leak that is non existent

echoz committed:
ivar for BOOL prettyTabBarButtons conflicts with internal ivar

echoz committed:
use ivar for accessor methods

echoz committed:
Merge branch 'master' of https://github.com/vicpenap/PrettyKit

echoz committed:
Merge branch 'master' into f6222169d60127a33b7789df688e99b7ca35d8fe

echoz committed:
Merge branch 'master' of github.com:echoz/PrettyKit

echoz committed:
added ability to generate iOS 6 styled tabbars

echoz committed:
include the correct framework

echoz committed:
Merge branch 'master' into PrettyKitOriginal
